src/sdam/description/server.rs

```diff
@@ -71,7 +91,7 @@ pub(crate) struct ServerDescription {
     // allows us to ensure that only valid states are possible (e.g. preventing that both an error
     // and a reply are present) while still making it easy to define helper methods on
     // ServerDescription for information we need from the isMaster reply by propagating with `?`.
-    pub(crate) reply: Result<Option<IsMasterReply>, String>,
+    pub(crate) reply: Result<Option<HelloReply>, String>,
```

Comment:
So in order to completely separate the implementation / public API completely, we introduced the `ServerInfo` struct (in `sdam::public`) a while back to be used for filtering in the `Predicate` selection criteria. I think we may want to also use that type for SDAM monitoring too to reduce duplication.

Given the naming requirements in the specification, we'd probably want to refactor `ServerInfo` to be named `ServerDescription` as part of that (the name won't clash internally since they're in different modules). One thing to consider with this is `ServerInfo` currently wraps a reference to our internal `ServerDescription`, and for the purposes of event emitting that will likely need to be updated to be an owned value. This may hurt the performance of `Predicate` based server selection though, so perhaps we would update `ServerInfo` to have a `Cow<'a, ServerDescription>` in it instead just a reference or just an owned value, and in the SDAM events that need it will just store a `ServerInfo<'static>`.

Similarly, instead of marking `TopologyDescription` public, we probably want to introduce a similar wrapper type into `sdam::public`.

Reply:
Bikeshed: I'd really prefer to not have two types with the same name, even if the compiler sees them as distinct.

Reply:
The wrappers would need to keep the internal field non-`pub` in order not to expose the current description types, right? My only concern here is that logging these events won't be super readable with the wrapper type + `Cow`.

(Also agree that duplicate naming can be somewhat confusing, although I'm not sure what we'd want to rename the internal `ServerDescription` type to.)